File: BlenderAgent/handlers/parametric.py
"""Parametric library dispatcher + proportion-matrix introspection.

Endpoints
---------
- /parametric/list                    — registered builders
- /parametric/build                   — invoke builder by name with params
- /parametric/anatomy/human           — resolve human canon to dims
- /parametric/anatomy/furniture       — defaults for known furniture kind
- /parametric/animation/timing        — frame timing for a known cycle
- /parametric/material/recipe         — PBR values for a material archetype
"""

from __future__ import annotations

import importlib
from typing import Any, Callable, Dict

from ..helpers import InvalidInputError, composite_undo
from ..library.anatomy import proportions as _props
from ..server import handler


# ---------------------------------------------------------------------------
# registry — each entry maps tool name → import path of the builder module
# ---------------------------------------------------------------------------

_BUILDERS: Dict[str, str] = {
    "chair_beach":   "BlenderAgent.library.furniture.chair_beach",
    "table_dining":  "BlenderAgent.library.furniture.table_dining",
    "stool_bar":     "BlenderAgent.library.furniture.stool_bar",
    "human_basic":   "BlenderAgent.library.character.human_basic",
}


def _resolve_builder(name: str) -> Callable[[Dict[str, Any]], Dict[str, Any]]:
    path = _BUILDERS.get(name)
    if path is None:
        raise InvalidInputError(
            f"unknown parametric builder {name!r}; valid: {sorted(_BUILDERS)}"
        )
    mod = importlib.import_module(path)
    build_fn = getattr(mod, "build", None)
    if not callable(build_fn):
        raise InvalidInputError(f"builder {name!r} has no `build` function")
    return build_fn


# ---------------------------------------------------------------------------
# /parametric/list
# ---------------------------------------------------------------------------

@handler("POST", "/parametric/list")
def parametric_list(_body: dict[str, Any]) -> dict[str, Any]:
    items = []
    for name in sorted(_BUILDERS):
        mod = importlib.import_module(_BUILDERS[name])
        doc = (getattr(mod, "build", None).__doc__ or "").strip()
        items.append({"name": name, "description": doc.splitlines()[0] if doc else ""})
    return {"ok": True, "data": {"builders": items, "count": len(items)}}


# ---------------------------------------------------------------------------
# /parametric/build
# ---------------------------------------------------------------------------

@handler("POST", "/parametric/build")
def parametric_build(body: dict[str, Any]) -> dict[str, Any]:
    """Body: {name: str, params?: dict}."""
    name = body.get("name")
    if not name:
        raise InvalidInputError("name is required")
    params = body.get("params") or {}
    if not isinstance(params, dict):
        raise InvalidInputError("params must be an object")

    build_fn = _resolve_builder(str(name))

    with composite_undo(f"parametric_build:{name}"):
        result = build_fn(params)

    parts = result.get("parts", {}) if isinstance(result, dict) else {}
    object_names = list(parts.values())
    return {
        "ok": True,
        "data": result,
        "refs": {"objectNames": object_names, "collectionName": result.get("collection")},
        "nextSteps": [
            "render a contact sheet via /vision/contact_sheet to verify silhouette",
            "run /vision/topology_inspect for quality gates",
        ],
    }


# ---------------------------------------------------------------------------
# anatomy / proportions introspection
# ---------------------------------------------------------------------------

@handler("POST", "/parametric/anatomy/human")
def parametric_anatomy_human(body: dict[str, Any]) -> dict[str, Any]:
    """Body: {canon?: str, height?: float, listCanons?: bool}.

    If listCanons is true, return the available canon names.
    Otherwise canon + height are required; returns derived dimensions.
    """
    if body.get("listCanons"):
        return {"ok": True, "data": {"canons": sorted(_props.HUMAN_CANONS.keys())}}
    canon = body.get("canon")
    height = body.get("height")
    if not canon or height is None:
        raise InvalidInputError("canon and height required (or listCanons=true)")
    try:
        dims = _props.resolve_human(str(canon), float(height))
    except ValueError as exc:
        raise InvalidInputError(str(exc)) from exc
    return {"ok": True, "data": dims}


@handler("POST", "/parametric/anatomy/furniture")
def parametric_anatomy_furniture(body: dict[str, Any]) -> dict[str, Any]:
    """Body: {kind?: str, list?: bool}."""
    if body.get("list"):
        return {"ok": True, "data": {"kinds": sorted(_props.FURNITURE_DEFAULTS.keys())}}
    kind = body.get("kind")
    if not kind:
        raise InvalidInputError("kind required (or list=true)")
    try:
        return {"ok": True, "data": _props.furniture_defaults(str(kind))}
    except ValueError as exc:
        raise InvalidInputError(str(exc)) from exc


@handler("POST", "/parametric/animation/timing")
def parametric_animation_timing(body: dict[str, Any]) -> dict[str, Any]:
    """Body: {kind?: str, fps?: float, list?: bool}."""
    if body.get("list"):
        return {"ok": True, "data": {"kinds": sorted(_props.ANIMATION_TIMINGS.keys())}}
    kind = body.get("kind")
    if not kind:
        raise InvalidInputError("kind required (or list=true)")
    fps = body.get("fps")
    try:
        return {"ok": True, "data": _props.animation_timing(str(kind), float(fps) if fps else None)}
    except ValueError as exc:
        raise InvalidInputError(str(exc)) from exc


@handler("POST", "/parametric/material/recipe")
def parametric_material_recipe(body: dict[str, Any]) -> dict[str, Any]:
    """Body: {name?: str, list?: bool}."""
    if body.get("list"):
        return {"ok": True, "data": {"recipes": sorted(_props.MATERIAL_RECIPES.keys())}}
    name = body.get("name")
    if not name:
        raise InvalidInputError("name required (or list=true)")
    try:
        return {"ok": True, "data": _props.material_recipe(str(name))}
    except ValueError as exc:
        raise InvalidInputError(str(exc)) from exc

File: Tools/src/tools/parametric.ts
/**
 * Parametric library tools (Sprint 6).
 *
 * Phase A of the aaa-modeling skill: prefer parametric builders over manual
 * primitive composition. Also exposes the proportion/timing/material matrices
 * (the "pair representation" from AlphaFold) for the agent to query before
 * making decisions.
 */

import { z } from "zod";
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";

import { passthroughPost, registerTools } from "../tool-helpers.js";

export function registerParametricTools(server: McpServer): void {
  registerTools(server, [
    {
      name: "parametric_list",
      description:
        "List every parametric builder registered in BlenderAgent/library. Each builder turns a small params dict into a topologically-clean named-parts asset.",
      inputSchema: {},
      handler: passthroughPost("/parametric/list"),
    },
    {
      name: "parametric_build",
      description:
        "Invoke a parametric builder by name. Pass partial params; missing keys fall back to the canonical defaults. Returns object names in refs for chaining into vision/postproc.",
      inputSchema: {
        name: z.string().describe(
          "Builder name (see parametric_list). E.g. 'chair_beach', 'table_dining', 'stool_bar'.",
        ),
        params: z
          .record(z.string(), z.unknown())
          .optional()
          .describe(
            "Builder-specific parameters. See the builder's docstring (parametric_list returns the first line of each). All keys are optional.",
          ),
      },
      handler: passthroughPost("/parametric/build"),
    },
    {
      name: "parametric_anatomy_human",
      description:
        "Resolve a human proportion canon (heroic_male, fashion, realistic_adult, chibi, etc.) at a given total height. Returns every landmark in meters. Use BEFORE creating an armature to lock proportions.",
      inputSchema: {
        canon: z.string().optional().describe(
          "Canon name. See listCanons. Required unless listCanons=true.",
        ),
        height: z.number().positive().optional().describe(
          "Total height in meters. Required unless listCanons=true.",
        ),
        listCanons: z.boolean().optional().describe(
          "List the available canon names instead of resolving.",
        ),
      },
      handler: passthroughPost("/parametric/anatomy/human"),
    },
    {
      name: "parametric_anatomy_furniture",
      description:
        "Return ergonomic default dimensions for a known furniture kind. Use BEFORE building to anchor the spec.",
      inputSchema: {
        kind: z.string().optional().describe(
          "Furniture kind. See list=true. Required unless list=true.",
        ),
        list: z.boolean().optional().describe("List known kinds instead."),
      },
      handler: passthroughPost("/parametric/anatomy/furniture"),
    },
    {
      name: "parametric_animation_timing",
      description:
        "Frame timing presets for common animation cycles (walk_cycle_human, run_cycle_human, idle_breathing, jump_neutral). Optionally rescale to a custom fps.",
      inputSchema: {
        kind: z.string().optional().describe(
          "Cycle name. See list=true.",
        ),
        fps: z.number().positive().optional().describe(
          "Target fps for rescaling (defaults to the preset's fps).",
        ),
        list: z.boolean().optional().describe("List known cycles instead."),
      },
      handler: passthroughPost("/parametric/animation/timing"),
    },
    {
      name: "parametric_material_recipe",
      description:
        "PBR values for a known material archetype (weathered_teak, polished_oak, faded_canvas_orange, brushed_aluminum, skin_caucasian, fabric_denim).",
      inputSchema: {
        name: z.string().optional().describe("Recipe name. See list=true."),
        list: z.boolean().optional().describe("List known recipes instead."),
      },
      handler: passthroughPost("/parametric/material/recipe"),
    },
  ]);
}
